tt_metal/impl/device/device_mesh_view.cpp

```diff
@@ -0,0 +1,194 @@
+
+#include "tt_metal/impl/device/device_mesh_view.hpp"
+#include <algorithm>
+#include <stdexcept>
+
+namespace tt::tt_metal {
+
+DeviceMeshView::DeviceMeshView(const DeviceMesh& mesh)
+    : top_left_(0, 0), bottom_right_(mesh.num_rows() - 1, mesh.num_cols() - 1) {
+    for (int row = 0; row < mesh.num_rows(); ++row) {
+        for (int col = 0; col < mesh.num_cols(); ++col) {
+            if (auto device = mesh.get_device(row, col)) {
+                devices_.push_back(device);
+                device_coordinates_[(device)->id()] = {row, col};
+            }
+        }
+    }
+}
+
+DeviceMeshView::DeviceMeshView(const DeviceMesh& mesh, Coordinate top_left, Coordinate bottom_right)
+    : top_left_(top_left), bottom_right_(bottom_right) {
+    for (int row = top_left.row; row <= bottom_right.row; ++row) {
+        for (int col = top_left.col; col <= bottom_right.col; ++col) {
+            if (auto device = mesh.get_device(row, col)) {
+                devices_.push_back(device);
+                device_coordinates_[(device)->id()] = {row, col};
+            }
+        }
+    }
+    validate_coordinates();
+}
+
+DeviceMeshView::DeviceMeshView(const DeviceMesh& global_mesh, const std::vector<device_pointer>& devices)
+    : devices_(devices) {
+    int min_row = std::numeric_limits<int>::max(), min_col = std::numeric_limits<int>::max();
+    int max_row = std::numeric_limits<int>::min(), max_col = std::numeric_limits<int>::min();
+
+    for (const auto& device : devices) {
+        if (auto coord = global_mesh.find_device(device->id())) {
+            device_coordinates_[device->id()] = *coord;
+            min_row = std::min(min_row, coord->row);
+            min_col = std::min(min_col, coord->col);
+            max_row = std::max(max_row, coord->row);
+            max_col = std::max(max_col, coord->col);
+        } else {
+            throw std::runtime_error("Device not found in global mesh");
+        }
+    }
+
+    top_left_ = {min_row, min_col};
+    bottom_right_ = {max_row, max_col};
+}
+
+DeviceMeshView::DeviceMeshView(std::vector<device_pointer> devices, CoordinateMapper mapper)
+    : devices_(std::move(devices)) {
+    initialize_from_devices(devices_, std::move(mapper));
+}
+
+DeviceMeshView::device_pointer DeviceMeshView::get_device(int row, int col) {
+    return const_cast<device_pointer>(std::as_const(*this).get_device(row, col));
+}
+
+DeviceMeshView::const_device_pointer DeviceMeshView::get_device(int row, int col) const {
+    for (const auto& device : devices_) {
+        auto it = device_coordinates_.find(device->id());
+        if (it != device_coordinates_.end() && it->second.row == row && it->second.col == col) {
+            return device;
+        }
+    }
+    return nullptr;
+}
+
+const std::vector<DeviceMeshView::device_pointer>& DeviceMeshView::get_devices() const {
+    return devices_;
+}
+
+std::vector<DeviceMeshView::device_pointer> DeviceMeshView::get_devices_on_row(int row) const {
```

Comment:
can you explain how this is different than `DeviceMesh::get_devices_on_row`?

Reply:
I've combined functionality now. thx!